File: python/cryptobi/toolbox/system/CBConfig.py
"""

/*
 *
 * Crypto.BI Toolbox
 * https://Crypto.BI/
 *
 * Author: José Fonseca (https://zefonseca.com/)
 *
 * Distributed under the MIT software license, see the accompanying
 * file COPYING or http://www.opensource.org/licenses/mit-license.php.
 *
 */

"""

from datetime import datetime
from pathlib import Path
import sys
import os
import configparser
import argparse

class CBConfig:

    """
        Crypto.BI Toolbox global configuration object.
        Use static method CBConfig.get_conf() to obtain a singleton.
    """

    # static consts
    DEFAULT_BITCOIN_DATADIR = ".bitcoin"
    DEFAULT_USER_DIR = ".Crypto.BI"
    DEFAULT_CONFIG_FILENAME = "crypto.bi.ini"
    DEFAULT_LOG_FILE = "crypto.bi.log"
    DEFAULT_LOG_ERROR_FILE = "crypto.bi.err"
    LOG_FILE_SIZE = 1048576 * 5
    LOG_FILE_DAYS = 3
    ENVIRONMENT_VAR_HOME = "CRYPTOBI_HOME"

    static_conf = None

    def __init__(self):
        self.cmd_parser = None
        self.config_map = {}
        self.init_config()

    @staticmethod
    def get_config():
        if CBConfig.static_conf == None:
            CBConfig.static_conf = CBConfig()
        return CBConfig.static_conf

    @staticmethod
    def get_user_dir():
        return str(Path.home())

    def read_from_file(self, config_path):
        cf = configparser.ConfigParser()
        cf.read(config_path)
        self.config_map.update(cf['DEFAULT'])

    def init_dir(self, config_dir) :
        if not os.path.isdir(config_dir):
            os.mkdir(config_dir)

    def get_conf(self, key):
        if key in self.config_map.keys():
            return self.config_map[key]
        return

    def apply_default_config(self):
        self.config_map["bitcoin_blocks_dir"] = os.path.join(CBConfig.get_user_dir(), CBConfig.DEFAULT_BITCOIN_DATADIR, "blocks")
        self.config_map["conf"] = os.path.join(CBConfig.get_user_dir(), CBConfig.DEFAULT_USER_DIR, CBConfig.DEFAULT_CONFIG_FILENAME)
        self.config_map["log_dir"] = os.path.join(CBConfig.get_user_dir(), CBConfig.DEFAULT_USER_DIR)

    def process_command_line(self, apply=True):
        parser = argparse.ArgumentParser()
        cryptobi_home = os.getenv(CBConfig.ENVIRONMENT_VAR_HOME)

        if cryptobi_home == None:
            self.log_error("Environment variable {} must be set in order to use Python Crypto.BI Toolbox".format(CBConfig.ENVIRONMENT_VAR_HOME))
            sys.exit(1)

        tsp = os.path.join(cryptobi_home,"toolbox","system","config_options.map")

        with open(tsp, 'r') as f:
            cft = f.read()
        params = "[" + cft + "]"
        param_list = eval(params.replace("\",","\":"))

        for ple in param_list:
            for k in ple.keys():
                v = ple[k]
                ak = "--" + k
                parser.add_argument(ak,help=v,type=str)

        parser.add_argument('listargs', type=str, nargs='?')
        args = vars(parser.parse_args())

        if (apply):
            for k in args:
                v = args[k]
                if (v is not None):
                    self.config_map[k] = v

    def log_message(self, message):
        print("[INFO] {} : {}".format(datetime.now(), message))

    def log_error(self, error):
        print("[ERROR] {} : {}".format(datetime.now(), error), file=sys.stderr)

    def init_config(self):
        self.apply_default_config()
        self.process_command_line(True)
        self.read_from_file(self.config_map["conf"])
        self.process_command_line(True)

File: toolbox/system/CBConfig.cpp
/*
 *
 * Crypto.BI Toolbox
 * https://Crypto.BI/
 *
 * Author: José Fonseca (https://zefonseca.com/)
 *
 * Distributed under the MIT software license, see the accompanying
 * file COPYING or http://www.opensource.org/licenses/mit-license.php.
 *
 */
#include "CBConfig.h"

namespace toolbox::system {

	std::shared_ptr<CBConfig> CBConfig::_config;

	CBConfig::CBConfig() {

		// initialize loggers
		console = spdlog::stdout_color_mt("console");
		err_logger = spdlog::stderr_color_mt("stderr");
		file_logger = spdlog::rotating_logger_mt("file_logger", DEFAULT_LOG_FILE , LOG_FILE_SIZE, LOG_FILE_DAYS);
		file_loggerr = spdlog::rotating_logger_mt("file_error_logger", DEFAULT_LOG_ERROR_FILE , LOG_FILE_SIZE, LOG_FILE_DAYS);

		// make sure user config dir exists
		boost::filesystem::path idir = get_user_dir() / boost::filesystem::path{DEFAULT_USER_DIR};
		bool id_ret = init_dir(idir);

		if (!id_ret) {
			log_error("init_dir() unable to initialize config directory.");
		}

		auto obx = desc.add_options();
		for (const auto &[k,v] : ___cb_opts) {
			obx(k.c_str(), po::value<std::string>(), v.c_str());
		}
	}

	CBConfig::~CBConfig() {

	}

	std::shared_ptr<CBConfig> CBConfig::get_config() {
		if (_config == nullptr) {
			_config = std::make_shared<CBConfig>();
		}
		return _config;
	}

	bool CBConfig::read_from_file(const boost::filesystem::path &config_path) {
		inipp::Ini<char> ini;
		std::ifstream is(config_path.generic_string());
		ini.parse(is);
		ini.interpolate();
		auto sym = ini.get_symbols();
		for (const auto &[name, value] : sym) {
			config_map[name.substr(3, name.length()-4)] = value;
		}
		return true;
	}

	bool CBConfig::init_dir(const boost::filesystem::path &config_dir) {

		if (boost::filesystem::exists(config_dir)) {
			return true;
		}

		return boost::filesystem::create_directory(config_dir);
	}

	std::optional<std::string> CBConfig::get_conf(const std::string &key) {
		try {
			auto ix = config_map.at(key);
			return ix;
		} catch (std::out_of_range &soor) {
			return {};
		}
	}

	/** Apply the correct order of parameter initialization.
	 *
	 * First, apply system defaults.
	 * Then, options file.
	 * Lastly the command line.
	 * Each step overwrites the previous, so the priority is applied
	 * in reverse, with the command line overriding the others.
	 *
	 * This is the function called by Toolbox programs' main().
	 *
	 * */
	bool CBConfig::init_config(int argc, char **argv) {
		auto _c = get_config(); // get or create config singleton
		apply_default_config(); // step 1
		// read options, see if config file changed
		// false indicates do not apply to config
		process_command_line(argc, argv, false);

		if (vm.count("conf") > 0) {
			_c->config_map["conf"] = vm["conf"].as<std::string>();
		}

		boost::filesystem::path cfile{_c->config_map["conf"]};

		if (boost::filesystem::exists(cfile)) {
			read_from_file(cfile); // step 2
		} else {
			log_error("Configuration file not found.");
		}

		// command line redone again, last step so it
		// overwrites all other config options
		process_command_line(argc, argv, true); // step 3

		// we're still here, noexcept. so, true
		return true;
	}

	boost::filesystem::path CBConfig::get_user_dir() {
		const char *homeDir = getenv("HOME");
		std::string hdir{homeDir};
		return boost::filesystem::path{hdir};
	}

	bool CBConfig::apply_default_config() {
		auto _c = get_config(); // get or create config singleton

		// set default blocks dir
		_c->config_map["bitcoin_blocks_dir"] = (
				get_user_dir() /
				boost::filesystem::path{DEFAULT_BITCOIN_DATADIR} /
				boost::filesystem::path{"blocks"}
				).generic_string();

		// set default config file
		_c->config_map["conf"] = (
				get_user_dir() /
				boost::filesystem::path{DEFAULT_USER_DIR} /
				boost::filesystem::path{DEFAULT_CONFIG_FILENAME}
				).generic_string();

		// set default log dir
		_c->config_map["log_dir"] = (
				get_user_dir() /
				boost::filesystem::path{DEFAULT_USER_DIR}
				).generic_string();

		return true;
	}

	/** Processes command line options.
	 * Only applies to overall config_map if apply is true.
	 *
	 * */
	bool CBConfig::process_command_line(int argc, char **argv, bool apply) {

		po::store(po::parse_command_line(argc, argv, desc), vm);
		po::notify(vm);

		// copy cmd line opts to internal config
		if (apply) {
			for (const auto &v : vm) {
				config_map[v.first] = v.second.as<std::string>();
			}
		}

		return true;
	}

	bool CBConfig::log_message(const std::string &message) {
	    spdlog::get("console")->info(message);
	    if (get_conf("log_to_file") == "1") {
	    	spdlog::get("file_logger")->info(message);
	    }
	    return true;
	}

	bool CBConfig::log_error(const std::string &error) {
		spdlog::get("stderr")->info(error);
	    if (get_conf("log_to_file") == "1") {
	    	spdlog::get("file_error_logger")->info(error);
	    }
		return true;
	}

} // namespace


